### PR TITLE
Verify the certificate chain against the Fulcio root trust by default

### PR DESCRIPTION
**Pull request by wata727**
#### Summary
See https://github.com/sigstore/cosign/issues/2134

Currently, the verification performed by `cosign verify-blob --cert` does not verify the certificate chain, so you cannot guarantee that the certificate was correctly issued by Fulcio. So you need to explicitly pass Fulcio's root certificate via the `--cert-chain` option.

```
cosign initialize
cat ~/.sigstore/root/targets/fulcio_intermediate_v1.crt.pem > fulcio_chained.crt.pem
echo >> fulcio_chained.crt.pem
cat ~/.sigstore/root/targets/fulcio.crt.pem >> fulcio_chained.crt.pem
echo >> fulcio_chained.crt.pem
cat ~/.sigstore/root/targets/fulcio_v1.crt.pem >> fulcio_chained.crt.pem

cosign verify-blob --cert checksum.txt.pem --cert-chanin fulcio_chained.crt.pem ...
```

However, this way of getting root certificates is not secure. Ideally, it should be retrieved by a TUF client. After discussing this matter in #2134, @haydentherapper says that there is room for changing the default behavior.

This PR will change the default behavior from not verifying the certificate chain if `--cert-chain` is not passed to verifying the certificate chain against the Fulcio root trust getting by a TUF client.

#### Release Note

- Changed the default behavior from not verifying the certificate chain if `--cert-chain` is not passed to verifying the certificate chain against the Fulcio root trust getting by a TUF client.

#### Documentation

- The description of the `--certificate` option needs to be changed.

### CODE REVIEW

**Comment by wata727** [comment]
I have two things to discuss:

- Should this change be followed by similar commands such as `cosign verify`?
- Is there a good way to add tests for this implementation? `verify_blob_test.go` didn't seem to have any tests for verification.

File: cmd/cosign/cli/verify/verify_blob.go
Comment (haydentherapper) [inline cmd/cosign/cli/verify/verify_blob.go:121]: We need to also update the documentation for the `--certificate` flag to note that it verifies the certificate against the Fulcio roots provided by TUF.

**Review by haydentherapper — CHANGES_REQUESTED** [review]
The same changes need to be made in https://cs.github.com/sigstore/cosign/blob/c5fda01a8ff33ca981f45a9f13e7fb6bd2080b94/cmd/cosign/cli/verify/verify_attestation.go?q=org%3Asigstore+CheckCertificatePolicy#L141 and https://cs.github.com/sigstore/cosign/blob/95b74db89941e8ec85e768f639efd4d948db06cd/cmd/cosign/cli/verify/verify.go?q=org%3Asigstore+CheckCertificatePolicy#L162

**Comment by haydentherapper** [comment]
Testing is unfortunately a bit rough. Check out `e2e_test.go`, but it might be hard to fit in a test for this, especially with TUF. We may need a TODO.

**Comment by haydentherapper** [comment]
/hold, I want to give a bit more thought to this and will get back to you on Monday.

cc @znewman01

File: cmd/cosign/cli/verify/verify_blob.go
Comment (wata727) [reply cmd/cosign/cli/verify/verify_blob.go:121]: Fixed https://github.com/sigstore/cosign/pull/2139/commits/b4887ea391a33cedd595d3b193baae5dc15d037e

**Comment by wata727** [comment]
> The same changes need to be made

Fixed https://github.com/sigstore/cosign/pull/2139/commits/eda267b9f4c323fdaa2f20330cff6ff137267b10

> Testing is unfortunately a bit rough. Check out e2e_test.go, but it might be hard to fit in a test for this, especially with TUF. We may need a TODO.

Understood. I'll take a little deeper look at the `e2e_test.go` implementation.

**Comment by znewman01** [comment]
Hey, thanks for this! (and thanks for looping me in, @haydentherapper)

I haven't looked too closely at the code but this change seems like much better/safer default behavior.

I just want to think through workarounds for anyone who was relying on the old behavior. The *right* thing to do in such a case is probably to provide the public key directly, rather than via a cert; this makes it much clearer what you're asking for. Maybe we should refer users to the [`step certificate key`](https://smallstep.com/docs/step-cli/reference/certificate/key) command? IIUC it should Just Work™

Was there anything specific you wanted my feedback on?

**Comment by patflynn** [comment]
I'm probably mis-interpreting this, but for cases where the user has signed a blob with a user provided custom key, why would we want to verify against the fulcio root?

**Comment by haydentherapper** [comment]
@znewman01 Just looking for a second opinion! Generally this looked good to me, beyond the potentially breaking change.

@patflynn One thought is a user is very unlikely to provide a certificate in that case, they'd be providing a key. We've got a few options for providing verification material:
* No key or cert - Either:
   * Fetch the certificate+chain from OCI. Verify the certificate+chain against the TUF root
   * Fetch the certificate from Rekor (for blob signing, and I don't actually like this since Rekor is meant to be a witness to an event and not storage, but let's discuss that more in another issue). 
* A key - Use the key to verify
* A certificate+chain - Verify the chain, and extract the verification key from the cert. Don't verify against TUF.
* A certificate w/o chain - Current behavior is to just extract the verification key. My question is, is anyone providing a certificate that's not from Fulcio in this case? If the vast majority of users provide a Fulcio certificate, then we should verify against the chain from TUF.

The workaround would either be extracting the key themselves (Like Zach mentioned) or providing the full certificate chain.

**Comment by patflynn** [comment]
@haydentherapper got it. that makes sense. Thanks for clarifying.

File: cmd/cosign/cli/options/certificate.go
Comment (asraa) [inline cmd/cosign/cli/options/certificate.go:40]: nit: Fulcio root provided in the TUF root?

**Review by asraa — APPROVED** [review]

**Comment by haydentherapper** [comment]
LGTM, just need to fix the failing test. Thanks!

**Comment by dlorenc** [comment]
The failure is unrelated and the fix is over here: https://github.com/sigstore/scaffolding/pull/277/

**Comment by dlorenc** [comment]
I'll merge while we wait for the fix.

File: cmd/cosign/cli/options/certificate.go
Comment (wata727) [reply cmd/cosign/cli/options/certificate.go:40]: I didn't mention it to TUF here because I have a concern about the description would be too long. If we should mention it, please let me know so I fix it.

**Pull request merged**
